### PR TITLE
Ropes Issue 3

### LINKED ISSUE #3: Image Index
#### As a user I want the homepage to display all saved images.

__Story__:

Now that we are saving images for our users, we want them to be able to view
the list of images in the system. You might be thinking, "what if my users only
want to view images of cats", or "what if my users only want to see the images
they uploaded". Those are great questions. Maybe you will have the opportunity
to add those features in a future story. For now, we will keep it simple and
quickly deliver a little more value to our users.

__Acceptance criteria__:
- [x] New images that are added show up on the homepage.
- [x] These images are persisted if the browser is closed or even if the
  server is restarted.
- [x] Images are not displayed wider than 400px.
- [x] Newest images appear first.

__Dependencies__:
- Save Image Link

### PR DESCRIPTION
**Pull request by gitlev**
Closes #3 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [x] I have replaced `X` above with the associated issue number.
- [x] I have changed the issue's label to "pull request".
- [x] I have assigned the designated reviewer to this PR.
- [x] I have set the appropriate label on this PR.
- [x] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: test/controllers/images_controller_test.rb [hunk 1]
```diff
@@ -6,6 +6,25 @@ class ImagesControllerTest < ActionController::TestCase
     assert_response :ok
   end
 
+  test 'index shows all images' do
+    Image.create!(url: 'http://www.factslides.com/imgs/Grizzly-bear.jpg', created_at: Time.new(2015).utc)
+    Image.create!(url: 'http://www.hallobay.com/wp-content/gallery/home/bear-close-up.jpg',
+                  created_at: Time.new(2016).utc)
```
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: Some fun facts about ruby/rails Time:
When you say `Time.new(2016).utc`, it takes `Time.new(2016)` in your set timezone and converts it to the UTC timezone, which is different than `Time.new(2016)` in the UTC timezone. You can play around with this in a rails console:
```
[7] pry(main)> Time.new(2016)
=> 2016-01-01 00:00:00 -0800
[8] pry(main)> Time.new(2016).utc
=> 2016-01-01 08:00:00 UTC
[9] pry(main)> Time.new(2016).utc == Time.new(2016)
=> true
[10] pry(main)> Time.utc(2016) == Time.new(2016).utc
=> false
```


Also, when times are saved to the database, they are converted from the timezone of the `Time` object, to the database timezone (probably UTC). So adding the `.utc` will actually be the same as not having the `.utc` when it is saved to the database.

Not a big deal either way, but it could be worth reading or playing around with ruby/rails `Time`.

File: test/controllers/images_controller_test.rb [hunk 2]
```diff
@@ -6,6 +6,25 @@ class ImagesControllerTest < ActionController::TestCase
...
+    get :index
+    assert_select 'img', 2
+  end
+
+  test 'index shows newest image first' do
+    Image.create!(url: 'http://www.factslides.com/imgs/Grizzly-bear.jpg', created_at: Time.new(2015).utc)
+    Image.create!(url: 'http://www.hallobay.com/wp-content/gallery/home/bear-close-up.jpg',
+                  created_at: Time.new(2016).utc)
+    get :index
```
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: What's the expected http response to this request?

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -6,6 +6,25 @@ class ImagesControllerTest < ActionController::TestCase
...
+    assert_select 'img' do |img_tags|
+      assert_equal 'http://www.hallobay.com/wp-content/gallery/home/bear-close-up.jpg',
+                   img_tags[0].attributes['src'].value
```
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: What happens if the application is just repeating the same image over and over? Would one of your tests fail?

File: test/controllers/images_controller_test.rb [hunk 4]
```diff
@@ -6,6 +6,25 @@ class ImagesControllerTest < ActionController::TestCase
...
+    end
+  end
+
```
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: What happens if there are no images in the database? Is it worth writing a test for this scenario?

File: Gemfile [hunk 5]
```diff
@@ -1,6 +1,7 @@
 source 'https://rubygems.org'
 ruby '2.3.2'
 
+gem 'hirb'
```
Comment (gitlev) [inline Gemfile]: makes console queries look nice

File: test/controllers/images_controller_test.rb (hunk 1 above)
Comment (gitlev) [inline test/controllers/images_controller_test.rb]: Good to know, your way is less convoluted, will change

File: test/controllers/images_controller_test.rb [hunk 6]
```diff
@@ -6,6 +6,25 @@ class ImagesControllerTest < ActionController::TestCase
     assert_response :ok
   end
 
+  test 'index shows all images' do
+    Image.create!(url: 'http://www.factslides.com/imgs/Grizzly-bear.jpg', created_at: Time.new(2015).utc)
+    Image.create!(url: 'http://www.hallobay.com/wp-content/gallery/home/bear-close-up.jpg',
+                  created_at: Time.new(2016).utc)
+    get :index
```
Comment (gitlev) [inline test/controllers/images_controller_test.rb]: It should return 200 ok but I test this in the 'should get index' test, so didn't include it in subsequent tests

File: test/controllers/images_controller_test.rb (hunk 2 above)
Comment (gitlev) [inline test/controllers/images_controller_test.rb]: It should return 200 ok but I test this in the 'should get index' test, so didn't include it in subsequent tests

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (gitlev) [inline test/controllers/images_controller_test.rb]: The test would still pass, but it would be trivial, but the url is just an identifier in my test to allow me to differentiate between the image created in 2015 and the image created in 2016, the image created in 2016 should be first on the page, and it's identifiable by the url by design in the test

File: test/controllers/images_controller_test.rb (hunk 4 above)
Comment (gitlev) [inline test/controllers/images_controller_test.rb]: The 'should get index' test covers this scenario, since there are no fixtures, the testing db is empty when it starts the test, and therefore the index page is rendered from a database with no images.  Do you think that testing that there are no images on the page in this case is worthwhile?

File: test/controllers/images_controller_test.rb (hunk 6 above)
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: It's generally a good idea to assert the response for every request, since different data could produce different results.

File: test/controllers/images_controller_test.rb (hunk 4 above)
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: Ahh, `should get index` was hidden in the diff, that should be good!

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (connorproctor) [inline test/controllers/images_controller_test.rb]: What I was trying to get at is that it could be a good idea to test that both urls are what you are expecting. Checking the url works. 

One other thing, it could be good to assign the `Image.create!` to a variable, say `most_recent_image`, and then use `most_recent_image.url` in the assertion. That way you aren't hardcoding the same url string twice
